lib/galaxy/jobs/__init__.py

```diff
@@ -1776,7 +1776,10 @@ def _finish_dataset(self, output_name, dataset, job, context, final_job_state, r
                 # call datatype.set_meta directly for the initial set_meta call during dataset creation
                 dataset.datatype.set_meta(dataset, overwrite=False)
             elif job.states.ERROR != final_job_state and not metadata_set_successfully:
-                dataset._state = model.Dataset.states.FAILED_METADATA
+                if self.tool.tool_type == "expression":
+                    dataset._state = model.Dataset.states.OK
+                else:
+                    dataset._state = model.Dataset.states.FAILED_METADATA
```

Comment:
Since `job.states.ERROR != final_job_state` is always True in the `elif not purged:` branch, I think the block starting from line 1768 to here can be rewritten as:

```python
            if not metadata_set_successfully:
                if self.tool.tool_type == "expression":
                    dataset._state = model.Dataset.states.OK
                elif retry_internally:
                    # If Galaxy was expected to sniff type and didn't - do so.
                    if dataset.ext == "_sniff_":
                        extension = sniff.handle_uploaded_dataset_file(
                            dataset.dataset.file_name, self.app.datatypes_registry
                        )
                        dataset.extension = extension

                    # call datatype.set_meta directly for the initial set_meta call during dataset creation
                    dataset.datatype.set_meta(dataset, overwrite=False)
                else:
                    dataset._state = model.Dataset.states.FAILED_METADATA
```

Reply:
Nice, that is a lot clearer!